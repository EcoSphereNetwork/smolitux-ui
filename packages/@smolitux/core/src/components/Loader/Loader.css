/* Loader Component Styles */
.smolitux-loader {
  /* Base loader styles */
}

.smolitux-loader__spinner {
  /* Spinner animation */
  animation: smolitux-spin 1s linear infinite;
}

.smolitux-loader__label {
  /* Label styles */
}

/* Variants */
.smolitux-loader--primary {
  /* Primary variant styles */
}

.smolitux-loader--subtle {
  /* Subtle variant styles */
}

.smolitux-loader--contrast {
  /* Contrast variant styles */
}

/* Spinner animation */
@keyframes smolitux-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}

/* Accessibility */
.sr-only {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border: 0;
}