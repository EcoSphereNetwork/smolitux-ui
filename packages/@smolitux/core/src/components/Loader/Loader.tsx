import React, { forwardRef } from 'react';
import { cn } from '@smolitux/utils';

export interface LoaderProps extends React.HTMLAttributes<HTMLDivElement> {
  /** Controls visibility of the loader */
  visible: boolean;
  /** Size of the loader */
  size?: 'sm' | 'md' | 'lg' | number;
  /** Label for screen readers or visible text */
  label?: string;
  /** Visual variant of the loader */
  variant?: 'primary' | 'subtle' | 'contrast';
  /** ARIA live region politeness */
  'aria-live'?: 'polite' | 'assertive';
  /** Fullscreen centered mode for loading screens */
  fullscreen?: boolean;
}

const Loader = forwardRef<HTMLDivElement, LoaderProps>(
  (
    {
      visible,
      size = 'md',
      label,
      variant = 'primary',
      'aria-live': ariaLive = 'polite',
      fullscreen = false,
      className,
      ...props
    },
    ref
  ) => {
    if (!visible) {
      return null;
    }

    const sizeValue = typeof size === 'number' ? size : getSizeValue(size);
    
    const loaderClasses = cn(
      'smolitux-loader',
      'inline-flex items-center justify-center',
      {
        'fixed inset-0 z-50 bg-white/80 backdrop-blur-sm': fullscreen,
        'smolitux-loader--primary': variant === 'primary',
        'smolitux-loader--subtle': variant === 'subtle',
        'smolitux-loader--contrast': variant === 'contrast',
      },
      className
    );

    const spinnerClasses = cn(
      'smolitux-loader__spinner',
      'animate-spin rounded-full border-2 border-current border-t-transparent',
      {
        'text-blue-600': variant === 'primary',
        'text-gray-400': variant === 'subtle',
        'text-white': variant === 'contrast',
      }
    );

    const labelClasses = cn(
      'smolitux-loader__label',
      'ml-2 text-sm font-medium',
      {
        'text-blue-600': variant === 'primary',
        'text-gray-600': variant === 'subtle',
        'text-white': variant === 'contrast',
      }
    );

    return (
      <div
        ref={ref}
        className={loaderClasses}
        role="status"
        aria-live={ariaLive}
        aria-label={label || 'Loading'}
        {...props}
      >
        <div className="flex items-center">
          <div
            className={spinnerClasses}
            style={{
              width: sizeValue,
              height: sizeValue,
            }}
            aria-hidden="true"
          />
          {label && (
            <span className={labelClasses}>
              {label}
            </span>
          )}
        </div>
        {!label && (
          <span className="sr-only">Loading...</span>
        )}
      </div>
    );
  }
);

function getSizeValue(size: 'sm' | 'md' | 'lg'): number {
  switch (size) {
    case 'sm':
      return 16;
    case 'md':
      return 24;
    case 'lg':
      return 32;
    default:
      return 24;
  }
}

Loader.displayName = 'Loader';

export default Loader;