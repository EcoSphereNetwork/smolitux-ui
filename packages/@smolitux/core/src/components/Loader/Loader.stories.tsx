import type { Meta, StoryObj } from '@storybook/react';
import Loader from './Loader';

const meta: Meta<typeof Loader> = {
  title: 'Core/Loader',
  component: Loader,
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component: 'A loading indicator component with customizable size, variant, and fullscreen mode.',
      },
    },
  },
  argTypes: {
    visible: {
      control: 'boolean',
      description: 'Controls visibility of the loader',
    },
    size: {
      control: { type: 'select' },
      options: ['sm', 'md', 'lg'],
      description: 'Size of the loader',
    },
    label: {
      control: 'text',
      description: 'Label for screen readers or visible text',
    },
    variant: {
      control: { type: 'select' },
      options: ['primary', 'subtle', 'contrast'],
      description: 'Visual variant of the loader',
    },
    'aria-live': {
      control: { type: 'select' },
      options: ['polite', 'assertive'],
      description: 'ARIA live region politeness',
    },
    fullscreen: {
      control: 'boolean',
      description: 'Fullscreen centered mode for loading screens',
    },
  },
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    visible: true,
    size: 'md',
    variant: 'primary',
  },
};

export const WithLabel: Story = {
  args: {
    visible: true,
    size: 'md',
    variant: 'primary',
    label: 'Loading content...',
  },
};

export const Small: Story = {
  args: {
    visible: true,
    size: 'sm',
    variant: 'primary',
    label: 'Loading...',
  },
};

export const Large: Story = {
  args: {
    visible: true,
    size: 'lg',
    variant: 'primary',
    label: 'Loading data...',
  },
};

export const CustomSize: Story = {
  args: {
    visible: true,
    size: 48,
    variant: 'primary',
    label: 'Custom size loader',
  },
};

export const Subtle: Story = {
  args: {
    visible: true,
    size: 'md',
    variant: 'subtle',
    label: 'Subtle loading...',
  },
};

export const Contrast: Story = {
  args: {
    visible: true,
    size: 'md',
    variant: 'contrast',
    label: 'Contrast loading...',
  },
  parameters: {
    backgrounds: { default: 'dark' },
  },
};

export const Invisible: Story = {
  args: {
    visible: false,
    size: 'md',
    variant: 'primary',
    label: 'This loader is not visible',
  },
};

export const Assertive: Story = {
  args: {
    visible: true,
    size: 'md',
    variant: 'primary',
    label: 'Important loading...',
    'aria-live': 'assertive',
  },
};

export const Fullscreen: Story = {
  args: {
    visible: true,
    size: 'lg',
    variant: 'primary',
    label: 'Loading application...',
    fullscreen: true,
  },
  parameters: {
    layout: 'fullscreen',
    docs: {
      story: {
        height: '400px',
      },
    },
  },
};

export const FullscreenContrast: Story = {
  args: {
    visible: true,
    size: 'lg',
    variant: 'contrast',
    label: 'Loading...',
    fullscreen: true,
  },
  parameters: {
    layout: 'fullscreen',
    backgrounds: { default: 'dark' },
    docs: {
      story: {
        height: '400px',
      },
    },
  },
};

// Interactive example showing visibility toggle
export const Interactive: Story = {
  args: {
    visible: true,
    size: 'md',
    variant: 'primary',
    label: 'Toggle visibility with the control',
  },
  parameters: {
    docs: {
      description: {
        story: 'Use the visible control to toggle the loader on and off.',
      },
    },
  },
};