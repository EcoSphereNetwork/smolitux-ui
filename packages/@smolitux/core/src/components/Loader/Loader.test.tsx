import React from 'react';
import { render, screen } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import Loader from './Loader';

expect.extend(toHaveNoViolations);

describe('Loader', () => {
  it('renders when visible is true', () => {
    render(<Loader visible={true} />);
    expect(screen.getByRole('status')).toBeInTheDocument();
  });

  it('does not render when visible is false', () => {
    render(<Loader visible={false} />);
    expect(screen.queryByRole('status')).not.toBeInTheDocument();
  });

  it('displays label when provided', () => {
    const label = 'Loading data...';
    render(<Loader visible={true} label={label} />);
    expect(screen.getByText(label)).toBeInTheDocument();
  });

  it('shows default loading text when no label provided', () => {
    render(<Loader visible={true} />);
    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });

  it('applies correct size classes', () => {
    const { rerender } = render(<Loader visible={true} size="sm" />);
    let spinner = document.querySelector('.smolitux-loader__spinner');
    expect(spinner).toHaveStyle({ width: '16px', height: '16px' });

    rerender(<Loader visible={true} size="md" />);
    spinner = document.querySelector('.smolitux-loader__spinner');
    expect(spinner).toHaveStyle({ width: '24px', height: '24px' });

    rerender(<Loader visible={true} size="lg" />);
    spinner = document.querySelector('.smolitux-loader__spinner');
    expect(spinner).toHaveStyle({ width: '32px', height: '32px' });
  });

  it('applies custom numeric size', () => {
    render(<Loader visible={true} size={40} />);
    const spinner = document.querySelector('.smolitux-loader__spinner');
    expect(spinner).toHaveStyle({ width: '40px', height: '40px' });
  });

  it('applies variant classes correctly', () => {
    const { rerender } = render(<Loader visible={true} variant="primary" />);
    let container = screen.getByRole('status');
    expect(container).toHaveClass('smolitux-loader--primary');

    rerender(<Loader visible={true} variant="subtle" />);
    container = screen.getByRole('status');
    expect(container).toHaveClass('smolitux-loader--subtle');

    rerender(<Loader visible={true} variant="contrast" />);
    container = screen.getByRole('status');
    expect(container).toHaveClass('smolitux-loader--contrast');
  });

  it('applies fullscreen mode correctly', () => {
    render(<Loader visible={true} fullscreen={true} />);
    const container = screen.getByRole('status');
    expect(container).toHaveClass('fixed', 'inset-0', 'z-50');
  });

  it('sets correct ARIA attributes', () => {
    render(<Loader visible={true} aria-live="assertive" label="Custom loading" />);
    const container = screen.getByRole('status');
    expect(container).toHaveAttribute('aria-live', 'assertive');
    expect(container).toHaveAttribute('aria-label', 'Custom loading');
  });

  it('sets default aria-live to polite', () => {
    render(<Loader visible={true} />);
    const container = screen.getByRole('status');
    expect(container).toHaveAttribute('aria-live', 'polite');
  });

  it('forwards ref correctly', () => {
    const ref = React.createRef<HTMLDivElement>();
    render(<Loader visible={true} ref={ref} />);
    expect(ref.current).toBeInstanceOf(HTMLDivElement);
  });

  it('applies custom className', () => {
    render(<Loader visible={true} className="custom-class" />);
    const container = screen.getByRole('status');
    expect(container).toHaveClass('custom-class');
  });

  it('passes through additional props', () => {
    render(<Loader visible={true} data-testid="loader-test" />);
    const container = screen.getByRole('status');
    expect(container).toHaveAttribute('data-testid', 'loader-test');
  });

  it('should have no accessibility violations', async () => {
    const { container } = render(<Loader visible={true} label="Loading content" />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should have no accessibility violations in fullscreen mode', async () => {
    const { container } = render(
      <Loader visible={true} fullscreen={true} label="Loading application" />
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('matches snapshot', () => {
    const { container } = render(
      <Loader 
        visible={true} 
        size="md" 
        variant="primary" 
        label="Loading..." 
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for fullscreen mode', () => {
    const { container } = render(
      <Loader 
        visible={true} 
        fullscreen={true}
        size="lg" 
        variant="contrast" 
        label="Loading application..." 
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});