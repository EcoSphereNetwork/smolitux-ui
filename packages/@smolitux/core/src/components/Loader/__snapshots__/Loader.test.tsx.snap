// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader matches snapshot 1`] = `
<div
  aria-label="Loading..."
  aria-live="polite"
  class="smolitux-loader inline-flex items-center justify-center smolitux-loader--primary"
  role="status"
>
  <div
    class="flex items-center"
  >
    <div
      aria-hidden="true"
      class="smolitux-loader__spinner animate-spin rounded-full border-2 border-current border-t-transparent text-blue-600"
      style="width: 24px; height: 24px;"
    />
    <span
      class="smolitux-loader__label ml-2 text-sm font-medium text-blue-600"
    >
      Loading...
    </span>
  </div>
</div>
`;

exports[`Loader matches snapshot for fullscreen mode 1`] = `
<div
  aria-label="Loading application..."
  aria-live="polite"
  class="smolitux-loader inline-flex items-center justify-center fixed inset-0 z-50 bg-white/80 backdrop-blur-sm smolitux-loader--contrast"
  role="status"
>
  <div
    class="flex items-center"
  >
    <div
      aria-hidden="true"
      class="smolitux-loader__spinner animate-spin rounded-full border-2 border-current border-t-transparent text-white"
      style="width: 32px; height: 32px;"
    />
    <span
      class="smolitux-loader__label ml-2 text-sm font-medium text-white"
    >
      Loading application...
    </span>
  </div>
</div>
`;
